neurodamus/node.py

```diff
@@ -1851,61 +1864,73 @@ def _coreneuron_restore(self):
 
-    # -
-    def _instantiate_simulation(self):
-        # Keep the initial RSS for the SHM file transfer calculations
-        self._initial_rss = SHMUtil.get_node_rss()
-        print_mem_usage()
-
-        self.load_targets()
-
-        # Check connection block configuration and raise warnings for overriding
-        # parameters
-        SimConfig.check_connections_configure(self._target_manager)
+    def compute_n_cycles(self):
+        """Determine the number of model-building cycles
 
-        # Check if user wants to build the model in several steps (only for CoreNeuron)
+        It is based on configuration and system constraints.
+        """
         n_cycles = SimConfig.modelbuilding_steps
-
-        # Without multi-cycle, it's a trivial model build. sub_targets is False
+        # No multi-cycle. Trivial result, this is always possible
         if n_cycles == 1:
-            self._build_model()
-            return
+            return n_cycles
 
         target = self._target_manager.get_target(self._target_spec)
         target_name = self._target_spec.name
-        cell_count = target.gid_count()
-        logging.info("Simulation target: %s, Cell count: %d", target_name, cell_count)
+        max_cell_count = target.max_gid_count_per_population()
+        logging.info(
+            "Simulation target: %s, Max cell count per population: %d", target_name, max_cell_count
+        )
 
-        if SimConfig.use_coreneuron and cell_count / n_cycles < MPI.size and cell_count > 0:
+        if SimConfig.use_coreneuron and max_cell_count / n_cycles < MPI.size and max_cell_count > 0:
             # coreneuron with no. ranks >> no. cells
-            # need to assign fake gids to artificial cells in empty threads during module building
-            # fake gids start from max_gid + 1
+            # need to assign fake gids to artificial cells in empty threads
+            # during module building fake gids start from max_gid + 1
             # currently not support engine plugin where target is loaded later
-            max_num_cycles = int(cell_count / MPI.size) or 1
+            # We can always have only 1 cycle. coreneuron throws an error if a
+            # rank does not have cells during a cycle. There is a way to prevent
+            # this for unbalanced multi-populations but if more than one cycle
+            # happens on a rank without instantiating cells another error raises.
+            # Thus, the number of cycles should be rounded down; on the safe side
+            max_num_cycles = int(max_cell_count / MPI.size) or 1
             if n_cycles > max_num_cycles:
                 logging.warning(
                     "Your simulation is using multi-cycle without enough cells.\n"
                     "  => Number of cycles has been automatically set to the max: %d",
                     max_num_cycles,
                 )
                 n_cycles = max_num_cycles
+        self._n_cycles = n_cycles
+        return None
```

Comment:
It might be better to still `return n_cycles` here, so that it is consistent with the Line 1857.  We can set `self._n_cycles = self.compute_n_cycles()` in `_build_model`.  What do you think ?

Reply:
ah yes, I did not notice line 1857